File: Python/Patterns/Singleton.py
# Create the Singleton pattern
import threading


class WeatherSingleton:
    # Create private class attributes
    _instance = None
    _instance_lock = threading.Lock()
    
    def __new__(cls):
        # Using the threading, lock the instance creation process
        with cls._instance_lock:
            # Make sure the instance doesn't already exist
            if cls._instance is None:
                # Create a new and ONLY instance
                cls._instance = super(WeatherSingleton, cls).__new__(cls)
                # Setup attributes now - we won't initialize again
                cls._instance._weather_data = "Sunny, 25°C" 
        return cls._instance
    
    @property
    def weather_data(self):
        return self._weather_data
    
    @weather_data.setter
    def weather_data(self, weather_data: str):
        self._weather_data = weather_data
        
# Create an original weather system
weather_system = WeatherSingleton()
print(weather_system.weather_data) # Expected output: "Cloudy, 50°C"

# Update the weather system and print new data
weather_system.weather_data = "Cloudy, 50°C"
print(weather_system.weather_data)  # Expected output: "Cloudy, 50°C"


new_system = WeatherSingleton()
print(new_system.weather_data)  # Expected output: "Cloudy, 50°C"

File: Typescript/Patterns/Singleton.ts
// Create a singleton
class WeatherSingleton {
    // Private: means this variable is not available anywhere but in this class
    // Static: means that this is accesssible by simply doing WeatherSingleton.instance
    private static instance; WeatherSingleton; // Holds the single instance
    private weatherData: string;

    // Constructor() is a special built-in class that gets called when trying to create a new Class()
    private constructor() {
        this.weatherData = "Sunny, 25C"
    }

    public static getInstance(): WeatherSingleton {
        // If we don't have a weather singleton, let's make one
        if (!WeatherSingleton.instance) {
            WeatherSingleton.instance = new WeatherSingleton();
        }

        // Otherwise, if we already have one... let's return it
        return WeatherSingleton.instance;
    }

    // Getter: this gets the current weather data, allows us to safely control HOW we return the data
    public getWeather(): string {
        return this.weatherData;
    }

    // Setter: this sets the current weather data, allows us to safely control HOW that happens
    public setWeather(newWeather: string): void {
        this.weatherData = newWeather;
    }
}

const system1 = WeatherSingleton.getInstance();
console.log(system1.getWeather());  // "Sunny, 25°C"

system1.setWeather("Cloudy, 18°C");

const system2 = WeatherSingleton.getInstance();
console.log(system2.getWeather());  // "Cloudy, 18°C" (same instance!)
